In [77]:
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.svm import SVC
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder

In [78]:
df1=pd.read_csv('/kaggle/input/house-price-prediction-batch6/train.csv')
df1

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     142          20       RL         78.0    11645   Pave   NaN      Reg   
1     365          60       RL          NaN    18800   Pave   NaN      IR1   
2     799          60       RL        104.0    13518   Pave   NaN      Reg   
3     976         160       FV          NaN     2651   Pave   NaN      Reg   
4    1158         120       RL         34.0     5001   Pave   NaN      IR1   
..    ...         ...      ...          ...      ...    ...   ...      ...   
929   782          60       RL         65.0     7153   Pave   NaN      Reg   
930   614          20       RL         70.0     8402   Pave   NaN      Reg   
931   878          60       RL         74.0     8834   Pave   NaN      Reg   
932  1037          20       RL         89.0    12898   Pave   NaN      IR1   
933   873          20       RL         74.0     8892   Pave   NaN      Reg   

    LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal  \
0           Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
1           Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
2           Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
3           Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
4           Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
..          ...       ...  ...      ...    ...   ...         ...     ...   
929         Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
930         Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
931         Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
932         HLS    AllPub  ...        0    NaN   NaN         NaN       0   
933         Lvl    AllPub  ...        0    NaN  GdWo         NaN       0   

    MoSold YrSold  SaleType  SaleCondition  SalePrice  
0        1   2006        WD         Normal     260000  
1        7   2006        WD         Normal     190000  
2        7   2009       New        Partial     485000  
3        4   2006        WD         Normal     165000  
4        7   2009        WD         Normal     230000  
..     ...    ...       ...            ...        ...  
929      7   2006        WD         Normal     175900  
930     12   2007       New        Partial     147000  
931      6   2009        WD         Normal     350000  
932      9   2009        WD         Normal     315500  
933     10   2008        WD         Normal     116000  

[934 rows x 81 columns]

In [79]:
df2=pd.read_csv('/kaggle/input/house-price-prediction-batch6/test.csv')
df2

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0      42          20       RL        115.0    16905   Pave   NaN      Reg   
1    1156          20       RL         90.0    10768   Pave   NaN      IR1   
2     239          20       RL         93.0    12030   Pave   NaN      Reg   
3     810          75       RM         90.0     8100   Pave   NaN      Reg   
4     471         120       RL          NaN     6820   Pave   NaN      IR1   
..    ...         ...      ...          ...      ...    ...   ...      ...   
229   389          20       RL         93.0     9382   Pave   NaN      IR1   
230  1193          50       RM         60.0     9600   Pave  Grvl      Reg   
231    95          60       RL         69.0     9337   Pave   NaN      IR1   
232  1171          80       RL         76.0     9880   Pave   NaN      Reg   
233   190         120       RL         41.0     4923   Pave   NaN      Reg   

    LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0           Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
1           Lvl    AllPub  ...         180        0    NaN    NaN         NaN   
2           Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
3           Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
4           Lvl    AllPub  ...         140        0    NaN    NaN         NaN   
..          ...       ...  ...         ...      ...    ...    ...         ...   
229         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
230         Lvl    AllPub  ...           0        0    NaN   GdWo         NaN   
231         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
232         Lvl    AllPub  ...           0      576     Gd  GdPrv         NaN   
233         Lvl    AllPub  ...         224        0    NaN    NaN         NaN   

    MiscVal MoSold  YrSold  SaleType  SaleCondition  
0         0      7    2007        WD         Normal  
1         0      7    2007        WD         Normal  
2         0     12    2007       New        Partial  
3         0     11    2009        WD         Normal  
4         0      6    2010        WD         Normal  
..      ...    ...     ...       ...            ...  
229       0      7    2008        WD         Normal  
230       0      7    2007        WD         Normal  
231       0      5    2007        WD         Normal  
232       0      7    2008        WD         Normal  
233       0      8    2008        WD         Normal  

[234 rows x 80 columns]

In [80]:
df2.shape

(234, 80)

In [81]:
df1.shape

(934, 81)

In [82]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 234 entries, 0 to 233
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             234 non-null    int64  
 1   MSSubClass     234 non-null    int64  
 2   MSZoning       234 non-null    object 
 3   LotFrontage    195 non-null    float64
 4   LotArea        234 non-null    int64  
 5   Street         234 non-null    object 
 6   Alley          15 non-null     object 
 7   LotShape       234 non-null    object 
 8   LandContour    234 non-null    object 
 9   Utilities      234 non-null    object 
 10  LotConfig      234 non-null    object 
 11  LandSlope      234 non-null    object 
 12  Neighborhood   234 non-null    object 
 13  Condition1     234 non-null    object 
 14  Condition2     234 non-null    object 
 15  BldgType       234 non-null    object 
 16  HouseStyle     234 non-null    object 
 17  OverallQual    234 non-null    int64  
 18  OverallCon

In [83]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 934 entries, 0 to 933
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             934 non-null    int64  
 1   MSSubClass     934 non-null    int64  
 2   MSZoning       934 non-null    object 
 3   LotFrontage    771 non-null    float64
 4   LotArea        934 non-null    int64  
 5   Street         934 non-null    object 
 6   Alley          56 non-null     object 
 7   LotShape       934 non-null    object 
 8   LandContour    934 non-null    object 
 9   Utilities      934 non-null    object 
 10  LotConfig      934 non-null    object 
 11  LandSlope      934 non-null    object 
 12  Neighborhood   934 non-null    object 
 13  Condition1     934 non-null    object 
 14  Condition2     934 non-null    object 
 15  BldgType       934 non-null    object 
 16  HouseStyle     934 non-null    object 
 17  OverallQual    934 non-null    int64  
 18  OverallCon

In [84]:
s = df1.isna().sum()
missing_values = s[s > 1]
print("shape of train data:", df1.shape ,"\nnumber of null in train data: ")
for col, num_missing in missing_values.items():
    print(f'{col}: {num_missing}')

shape of train data: (934, 81) 
number of null in train data: 
LotFrontage: 163
Alley: 878
MasVnrType: 542
MasVnrArea: 6
BsmtQual: 26
BsmtCond: 26
BsmtExposure: 26
BsmtFinType1: 26
BsmtFinType2: 27
FireplaceQu: 436
GarageType: 50
GarageYrBlt: 50
GarageFinish: 50
GarageQual: 50
GarageCond: 50
PoolQC: 931
Fence: 755
MiscFeature: 898


In [85]:
s = df2.isna().sum()
missing_values = s[s > 1]
print("shape of test data:", df1.shape ,"\nnumber of null in test data: ")
for col, num_missing in missing_values.items():
    print(f'{col}: {num_missing}')

shape of test data: (934, 81) 
number of null in test data: 
LotFrontage: 39
Alley: 219
MasVnrType: 148
BsmtQual: 6
BsmtCond: 6
BsmtExposure: 6
BsmtFinType1: 6
BsmtFinType2: 6
FireplaceQu: 110
GarageType: 19
GarageYrBlt: 19
GarageFinish: 19
GarageQual: 19
GarageCond: 19
PoolQC: 232
Fence: 187
MiscFeature: 226


In [86]:
df1.describe()

Id  MSSubClass  LotFrontage        LotArea  OverallQual  \
count   934.000000  934.000000   771.000000     934.000000   934.000000   
mean    720.086724   57.692719    69.584955   10456.070664     6.139186   
std     425.245539   43.581647    22.718869   10533.263205     1.385852   
min       1.000000   20.000000    21.000000    1300.000000     1.000000   
25%     353.250000   20.000000    59.000000    7500.000000     5.000000   
50%     705.500000   50.000000    70.000000    9552.000000     6.000000   
75%    1101.500000   70.000000    80.000000   11644.500000     7.000000   
max    1459.000000  190.000000   182.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count   934.000000   934.000000    934.000000   928.000000   934.000000  ...   
mean      5.552463  1972.568522   1985.350107   107.657328   457.522484  ...   
std       1.119318    30.060124     20.480156   184.543020   444.230598  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1955.000000   1968.000000     0.000000     0.000000  ...   
50%       5.000000  1976.000000   1994.500000     0.000000   403.000000  ...   
75%       6.000000  2001.000000   2004.000000   170.000000   735.000000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  2260.000000  ...   

       WoodDeckSF  OpenPorchSF  EnclosedPorch   3SsnPorch  ScreenPorch  \
count  934.000000   934.000000     934.000000  934.000000   934.000000   
mean    99.177730    48.443255      21.120985    3.701285    13.375803   
std    124.777609    68.426148      60.321125   32.129822    52.363548   
min      0.000000     0.000000       0.000000    0.000000     0.000000   
25%      0.000000     0.000000       0.000000    0.000000     0.000000   
50%     29.000000    27.000000       0.000000    0.000000     0.000000   
75%    172.750000    70.000000       0.000000    0.000000     0.000000   
max    857.000000   547.000000     552.000000  508.000000   480.000000   

         PoolArea       MiscVal      MoSold       YrSold      SalePrice  
count  934.000000    934.000000  934.000000   934.000000     934.000000  
mean     1.932548     53.276231    6.332976  2007.814775  182981.457173  
std     34.512758    606.248612    2.711067     1.334253   80613.726628  
min      0.000000      0.000000    1.000000  2006.000000   34900.000000  
25%      0.000000      0.000000    5.000000  2007.000000  132000.000000  
50%      0.000000      0.000000    6.000000  2008.000000  165075.000000  
75%      0.000000      0.000000    8.000000  2009.000000  215000.000000  
max    738.000000  15500.000000   12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [87]:
df1.describe(include=object)

MSZoning Street Alley LotShape LandContour Utilities LotConfig  \
count       934    934    56      934         934       934       934   
unique        5      2     2        4           4         1         5   
top          RL   Pave  Grvl      Reg         Lvl    AllPub    Inside   
freq        740    930    29      579         849       934       655   

       LandSlope Neighborhood Condition1  ... GarageType GarageFinish  \
count        934          934        934  ...        884          884   
unique         3           25          9  ...          6            3   
top          Gtl        NAmes       Norm  ...     Attchd          Unf   
freq         896          153        811  ...        572          370   

       GarageQual GarageCond PavedDrive PoolQC  Fence MiscFeature SaleType  \
count         884        884        934      3    179          36      934   
unique          4          5          3      2      4           3        9   
top            TA         TA          Y     Ex  MnPrv        Shed       WD   
freq          848        855        866      2     91          32      815   

       SaleCondition  
count            934  
unique             6  
top           Normal  
freq             774  

[4 rows x 43 columns]

In [88]:
df2.describe()

Id  MSSubClass  LotFrontage        LotArea  OverallQual  \
count   234.000000  234.000000   195.000000     234.000000   234.000000   
mean    759.782051   56.111111    71.030769   10821.769231     5.944444   
std     424.277360   43.355062    33.047593   11522.432087     1.411598   
min      12.000000   20.000000    21.000000    1491.000000     2.000000   
25%     378.500000   20.000000    59.000000    7514.500000     5.000000   
50%     780.000000   50.000000    65.000000    9338.500000     6.000000   
75%    1140.250000   60.000000    80.000000   11357.750000     7.000000   
max    1460.000000  190.000000   313.000000  159000.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd  MasVnrArea   BsmtFinSF1  ...  \
count   234.000000   234.000000    234.000000  234.000000   234.000000  ...   
mean      5.666667  1969.534188   1986.081197   90.102564   416.170940  ...   
std       1.119473    30.268486     20.416212  160.431841   536.324925  ...   
min       2.000000  1893.000000   1950.000000    0.000000     0.000000  ...   
25%       5.000000  1951.250000   1969.000000    0.000000     0.000000  ...   
50%       5.000000  1970.500000   1994.500000    0.000000   333.000000  ...   
75%       6.000000  1998.750000   2004.750000  136.500000   670.250000  ...   
max       9.000000  2009.000000   2009.000000  796.000000  5644.000000  ...   

        GarageArea  WoodDeckSF  OpenPorchSF  EnclosedPorch   3SsnPorch  \
count   234.000000  234.000000    234.00000     234.000000  234.000000   
mean    456.735043   96.128205     45.67094      20.290598    5.782051   
std     227.767308  136.144019     66.33145      58.763666   33.339835   
min       0.000000    0.000000      0.00000       0.000000    0.000000   
25%     308.000000    0.000000      0.00000       0.000000    0.000000   
50%     461.000000    0.000000     24.50000       0.000000    0.000000   
75%     576.000000  168.000000     67.50000       0.000000    0.000000   
max    1418.000000  736.000000    502.00000     294.000000  290.000000   

       ScreenPorch    PoolArea      MiscVal      MoSold       YrSold  
count   234.000000  234.000000   234.000000  234.000000   234.000000  
mean     15.619658    4.512821    30.341880    6.235043  2007.816239  
std      54.861634   48.911344   189.584735    2.635245     1.347543  
min       0.000000    0.000000     0.000000    1.000000  2006.000000  
25%       0.000000    0.000000     0.000000    4.250000  2007.000000  
50%       0.000000    0.000000     0.000000    6.000000  2008.000000  
75%       0.000000    0.000000     0.000000    8.000000  2009.000000  
max     396.000000  576.000000  2000.000000   12.000000  2010.000000  

[8 rows x 37 columns]

In [89]:
df2.describe(include=object)

MSZoning Street Alley LotShape LandContour Utilities LotConfig  \
count       234    234    15      234         234       234       234   
unique        4      1     2        4           4         2         4   
top          RL   Pave  Grvl      Reg         Lvl    AllPub    Inside   
freq        187    234     9      152         210       233       172   

       LandSlope Neighborhood Condition1  ... GarageType GarageFinish  \
count        234          234        234  ...        215          215   
unique         3           24          7  ...          5            3   
top          Gtl        NAmes       Norm  ...     Attchd          Unf   
freq         219           32        197  ...        126          103   

       GarageQual GarageCond PavedDrive PoolQC  Fence MiscFeature SaleType  \
count         215        215        234      2     47           8      234   
unique          4          4          3      1      4           1        6   
top            TA         TA          Y     Gd  MnPrv        Shed       WD   
freq          196        201        212      2     32           8      200   

       SaleCondition  
count            234  
unique             6  
top           Normal  
freq             189  

[4 rows x 43 columns]

In [90]:
cat_feat = []
nue_feat = []
col_types = []

In [91]:
for col in df1.columns:
    col_types.append(df1[col].dtype)

In [92]:
pd.Series(col_types).unique()

array([dtype('int64'), dtype('O'), dtype('float64')], dtype=object)

In [93]:
for col in df1.columns:
    if 'object' in str(df1[col].dtype):
        cat_feat.append(col)
    else: nue_feat.append(col)

In [94]:
cat_feat

['MSZoning',
 'Street',
 'Alley',
 'LotShape',
 'LandContour',
 'Utilities',
 'LotConfig',
 'LandSlope',
 'Neighborhood',
 'Condition1',
 'Condition2',
 'BldgType',
 'HouseStyle',
 'RoofStyle',
 'RoofMatl',
 'Exterior1st',
 'Exterior2nd',
 'MasVnrType',
 'ExterQual',
 'ExterCond',
 'Foundation',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Heating',
 'HeatingQC',
 'CentralAir',
 'Electrical',
 'KitchenQual',
 'Functional',
 'FireplaceQu',
 'GarageType',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PavedDrive',
 'PoolQC',
 'Fence',
 'MiscFeature',
 'SaleType',
 'SaleCondition']

In [95]:
'''cat_feat = df1[cat_feat]
nue_feat = df1[nue_feat]'''

'cat_feat = df1[cat_feat]\nnue_feat = df1[nue_feat]'

In [96]:
#cat_feat.shape, nue_feat.shape

In [97]:
cat_feat

['MSZoning',
 'Street',
 'Alley',
 'LotShape',
 'LandContour',
 'Utilities',
 'LotConfig',
 'LandSlope',
 'Neighborhood',
 'Condition1',
 'Condition2',
 'BldgType',
 'HouseStyle',
 'RoofStyle',
 'RoofMatl',
 'Exterior1st',
 'Exterior2nd',
 'MasVnrType',
 'ExterQual',
 'ExterCond',
 'Foundation',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Heating',
 'HeatingQC',
 'CentralAir',
 'Electrical',
 'KitchenQual',
 'Functional',
 'FireplaceQu',
 'GarageType',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PavedDrive',
 'PoolQC',
 'Fence',
 'MiscFeature',
 'SaleType',
 'SaleCondition']

# see the corr of cat feature and target

In [98]:
def correlation_ratio(categories, measurements):
    fcat, _ = pd.factorize(categories)
    cat_num = np.max(fcat) + 1
    y_avg_array = np.zeros(cat_num)
    n_array = np.zeros(cat_num)
    
    for i in range(0, cat_num):
        cat_measures = measurements[np.argwhere(fcat == i).flatten()]
        n_array[i] = len(cat_measures)
        y_avg_array[i] = np.average(cat_measures)
    
    y_total_avg = np.sum(np.multiply(y_avg_array, n_array)) / np.sum(n_array)
    
    numerator = np.sum(np.multiply(n_array, (y_avg_array - y_total_avg) ** 2))
    denominator = np.sum((measurements - y_total_avg) ** 2)
    
    if numerator == 0:
        return 0.0
    
    return np.sqrt(numerator / denominator)

def correlation_ratios(data, categorical_features, target_feature):
    correlations = {}
    for cat_feature in categorical_features:
        ratio = correlation_ratio(data[cat_feature], data[target_feature])
        correlations[cat_feature] = ratio
    return correlations


target_feature = 'SalePrice' 


correlations = correlation_ratios(df1,cat_feat,target_feature)
sorted_correlations = sorted(correlations.items(), key=lambda item: item[1], reverse=True)

for cat_feature, ratio in sorted_correlations:
    print(f"{cat_feature}: {ratio}")


Neighborhood: 0.7241911234293448
KitchenQual: 0.6783575663504264
ExterQual: 0.6745851410603996
BsmtQual: 0.6633526555559756
Foundation: 0.5095458313146748
GarageFinish: 0.4680233532368886
BsmtFinType1: 0.4477673885808085
HeatingQC: 0.4384719014992851
Exterior2nd: 0.40788260221638784
GarageType: 0.3972637952481982
Exterior1st: 0.374763698717642
SaleType: 0.3631197306525389
BsmtExposure: 0.3573876569257375
SaleCondition: 0.34906217315143157
MSZoning: 0.30892450606188226
RoofStyle: 0.26435679820973385
HouseStyle: 0.2633784396100664
LotShape: 0.263170892723789
CentralAir: 0.2584665999076131
FireplaceQu: 0.25183066178864083
RoofMatl: 0.23915735336343386
Electrical: 0.2382837552903878
PavedDrive: 0.22467358813568108
MasVnrType: 0.22241775202708178
Condition1: 0.18025665347156986
BldgType: 0.17420960653160872
BsmtCond: 0.16910895048210306
ExterCond: 0.15429358139099517
LandContour: 0.1430573698707933
GarageCond: 0.13519892782190157
Heating: 0.13318471991263559
LotConfig: 0.12999751120480457
G

# see the corr of int feature and target

In [99]:
# Select only numeric columns for correlation
numeric_data = df1.select_dtypes(include=[float, int])

# Handle missing values by filling them with the median of the column
numeric_data = numeric_data.fillna(numeric_data.median())

# Correlation matrix
correlation_matrix = numeric_data.corr()

# Extract the correlations with SalePrice and sort them
correlation_with_saleprice = correlation_matrix['SalePrice'].sort_values(ascending=False)

print(correlation_with_saleprice)

SalePrice        1.000000
OverallQual      0.777172
GrLivArea        0.720202
TotalBsmtSF      0.658005
1stFlrSF         0.643107
GarageCars       0.633413
GarageArea       0.617788
FullBath         0.547057
TotRmsAbvGrd     0.522824
YearBuilt        0.506217
YearRemodAdd     0.486246
MasVnrArea       0.473106
Fireplaces       0.472682
GarageYrBlt      0.445172
BsmtFinSF1       0.414819
LotFrontage      0.389367
OpenPorchSF      0.303031
2ndFlrSF         0.297547
WoodDeckSF       0.289732
LotArea          0.263302
HalfBath         0.246890
BsmtUnfSF        0.225948
BsmtFullBath     0.221758
BedroomAbvGr     0.157828
PoolArea         0.156577
ScreenPorch      0.102507
3SsnPorch        0.054778
MoSold           0.034923
Id               0.007412
BsmtHalfBath    -0.004336
BsmtFinSF2      -0.015181
LowQualFinSF    -0.015383
YrSold          -0.019660
MiscVal         -0.023140
OverallCond     -0.075046
MSSubClass      -0.091493
EnclosedPorch   -0.118967
KitchenAbvGr    -0.134543
Name: SalePr

# see if there a dup on data 

In [100]:
#edit
columns_with_duplicates = []

for col in df1.columns:
    if df1[col].duplicated().any():
        columns_with_duplicates.append(col)

# Print the columns with duplicates
print("Columns with duplicates:")
for col in columns_with_duplicates:
    print(col)

Columns with duplicates:
MSSubClass
MSZoning
LotFrontage
LotArea
Street
Alley
LotShape
LandContour
Utilities
LotConfig
LandSlope
Neighborhood
Condition1
Condition2
BldgType
HouseStyle
OverallQual
OverallCond
YearBuilt
YearRemodAdd
RoofStyle
RoofMatl
Exterior1st
Exterior2nd
MasVnrType
MasVnrArea
ExterQual
ExterCond
Foundation
BsmtQual
BsmtCond
BsmtExposure
BsmtFinType1
BsmtFinSF1
BsmtFinType2
BsmtFinSF2
BsmtUnfSF
TotalBsmtSF
Heating
HeatingQC
CentralAir
Electrical
1stFlrSF
2ndFlrSF
LowQualFinSF
GrLivArea
BsmtFullBath
BsmtHalfBath
FullBath
HalfBath
BedroomAbvGr
KitchenAbvGr
KitchenQual
TotRmsAbvGrd
Functional
Fireplaces
FireplaceQu
GarageType
GarageYrBlt
GarageFinish
GarageCars
GarageArea
GarageQual
GarageCond
PavedDrive
WoodDeckSF
OpenPorchSF
EnclosedPorch
3SsnPorch
ScreenPorch
PoolArea
PoolQC
Fence
MiscFeature
MiscVal
MoSold
YrSold
SaleType
SaleCondition
SalePrice


# full null in numeric feature in train data

In [101]:
df1['LotFrontage'] = df1['LotFrontage'].fillna(df1['LotFrontage'].median())
df1['MasVnrArea'] = df1['MasVnrArea'].fillna(df1['MasVnrArea'].median())
df1['GarageYrBlt'] = df1['GarageYrBlt'].fillna(df1['GarageYrBlt'].median())

# full null in cat feature in train data

In [102]:
# Fill categorical features with 'None'
df1['Alley'] = df1['Alley'].fillna('no')
df1['BsmtQual'] = df1['BsmtQual'].fillna('no')
df1['BsmtCond'] = df1['BsmtCond'].fillna('no')
df1['BsmtExposure'] = df1['BsmtExposure'].fillna('no')
df1['BsmtFinType1'] = df1['BsmtFinType1'].fillna('no')
df1['BsmtFinType2'] = df1['BsmtFinType2'].fillna('no')
df1['FireplaceQu'] = df1['FireplaceQu'].fillna('no')
df1['GarageType'] = df1['GarageType'].fillna('no')
df1['GarageFinish'] = df1['GarageFinish'].fillna('no')
df1['GarageQual'] = df1['GarageQual'].fillna('no')
df1['GarageCond'] = df1['GarageCond'].fillna('no')
df1['PoolQC'] = df1['PoolQC'].fillna('no')
df1['Fence'] = df1['Fence'].fillna('no')
df1['MiscFeature'] = df1['MiscFeature'].fillna('no')
df1['MasVnrType'] = df1['MasVnrType'].fillna('no')

shape of test data: (934, 81) 
number of null in test data: 
Alley: 219
MasVnrType: 148
BsmtQual: 6
BsmtCond: 6
BsmtExposure: 6
BsmtFinType1: 6
BsmtFinType2: 6
FireplaceQu: 110
GarageType: 19
GarageFinish: 19
GarageQual: 19
GarageCond: 19
PoolQC: 232
Fence: 187
MiscFeature: 226

In [103]:

df1['Electrical'] = df1['Electrical'].fillna(df1['Electrical'].mode()[0])


# full null in test data

In [104]:
# Fill categorical features with 'None'
df2['Alley'] = df2['Alley'].fillna('no')
df2['BsmtQual'] = df2['BsmtQual'].fillna('no')
df2['BsmtCond'] = df2['BsmtCond'].fillna('no')
df2['BsmtExposure'] = df2['BsmtExposure'].fillna('no')
df2['BsmtFinType1'] = df2['BsmtFinType1'].fillna('no')
df2['BsmtFinType2'] = df1['BsmtFinType2'].fillna('no')
df2['FireplaceQu'] = df1['FireplaceQu'].fillna('no')
df2['GarageType'] = df2['GarageType'].fillna('no')
df2['GarageFinish'] = df2['GarageFinish'].fillna('no')
df2['GarageQual'] = df2['GarageQual'].fillna('no')
df2['GarageCond'] = df2['GarageCond'].fillna('no')
df2['PoolQC'] = df2['PoolQC'].fillna('no')
df2['Fence'] = df2['Fence'].fillna('no')
df2['MiscFeature'] = df2['MiscFeature'].fillna('no')
df2['MasVnrType'] = df2['MasVnrType'].fillna('no')

In [105]:
df2['LotFrontage'] = df2['LotFrontage'].fillna(df2['LotFrontage'].median())
df2['GarageYrBlt'] = df2['GarageYrBlt'].fillna(df2['GarageYrBlt'].median())

In [106]:

df2['Electrical'] = df2['Electrical'].fillna(df2['Electrical'].mode()[0])


In [107]:
z=df1['SalePrice']
df1=df1.drop(columns='SalePrice')

In [108]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 934 entries, 0 to 933
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             934 non-null    int64  
 1   MSSubClass     934 non-null    int64  
 2   MSZoning       934 non-null    object 
 3   LotFrontage    934 non-null    float64
 4   LotArea        934 non-null    int64  
 5   Street         934 non-null    object 
 6   Alley          934 non-null    object 
 7   LotShape       934 non-null    object 
 8   LandContour    934 non-null    object 
 9   Utilities      934 non-null    object 
 10  LotConfig      934 non-null    object 
 11  LandSlope      934 non-null    object 
 12  Neighborhood   934 non-null    object 
 13  Condition1     934 non-null    object 
 14  Condition2     934 non-null    object 
 15  BldgType       934 non-null    object 
 16  HouseStyle     934 non-null    object 
 17  OverallQual    934 non-null    int64  
 18  OverallCon

In [109]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 234 entries, 0 to 233
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             234 non-null    int64  
 1   MSSubClass     234 non-null    int64  
 2   MSZoning       234 non-null    object 
 3   LotFrontage    234 non-null    float64
 4   LotArea        234 non-null    int64  
 5   Street         234 non-null    object 
 6   Alley          234 non-null    object 
 7   LotShape       234 non-null    object 
 8   LandContour    234 non-null    object 
 9   Utilities      234 non-null    object 
 10  LotConfig      234 non-null    object 
 11  LandSlope      234 non-null    object 
 12  Neighborhood   234 non-null    object 
 13  Condition1     234 non-null    object 
 14  Condition2     234 non-null    object 
 15  BldgType       234 non-null    object 
 16  HouseStyle     234 non-null    object 
 17  OverallQual    234 non-null    int64  
 18  OverallCon

# now we will se the outlier

In [110]:
def count_outliers(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = df[(df[column] < lower_bound) | (df[column] > upper_bound)]
    return len(outliers)
outlier_count = {}
for column in df1.select_dtypes(include='int'):
    outlier_count[column] = count_outliers(df1, column)
for column, count in outlier_count.items():
    print(f'{column}: {count} outliers')

Id: 0 outliers
MSSubClass: 72 outliers
LotArea: 37 outliers
OverallQual: 2 outliers
OverallCond: 82 outliers
YearBuilt: 7 outliers
YearRemodAdd: 0 outliers
BsmtFinSF1: 4 outliers
BsmtFinSF2: 100 outliers
BsmtUnfSF: 25 outliers
TotalBsmtSF: 40 outliers
1stFlrSF: 10 outliers
2ndFlrSF: 2 outliers
LowQualFinSF: 14 outliers
GrLivArea: 18 outliers
BsmtFullBath: 0 outliers
BsmtHalfBath: 54 outliers
FullBath: 0 outliers
HalfBath: 0 outliers
BedroomAbvGr: 24 outliers
KitchenAbvGr: 41 outliers
TotRmsAbvGrd: 16 outliers
Fireplaces: 1 outliers
GarageCars: 3 outliers
GarageArea: 14 outliers
WoodDeckSF: 16 outliers
OpenPorchSF: 53 outliers
EnclosedPorch: 128 outliers
3SsnPorch: 15 outliers
ScreenPorch: 67 outliers
PoolArea: 3 outliers
MiscVal: 35 outliers
MoSold: 0 outliers
YrSold: 0 outliers


In [111]:
'''for col in df1.select_dtypes(include='int'):
    plt.figure(figsize=(6, 4))
    plt.boxplot(df1[col], vert=False)
    plt.title(f'Box Plot for {col}')
    plt.xlabel('Values')
    plt.show()'''

"for col in df1.select_dtypes(include='int'):\n    plt.figure(figsize=(6, 4))\n    plt.boxplot(df1[col], vert=False)\n    plt.title(f'Box Plot for {col}')\n    plt.xlabel('Values')\n    plt.show()"

In [112]:
outlier_count

{'Id': 0,
 'MSSubClass': 72,
 'LotArea': 37,
 'OverallQual': 2,
 'OverallCond': 82,
 'YearBuilt': 7,
 'YearRemodAdd': 0,
 'BsmtFinSF1': 4,
 'BsmtFinSF2': 100,
 'BsmtUnfSF': 25,
 'TotalBsmtSF': 40,
 '1stFlrSF': 10,
 '2ndFlrSF': 2,
 'LowQualFinSF': 14,
 'GrLivArea': 18,
 'BsmtFullBath': 0,
 'BsmtHalfBath': 54,
 'FullBath': 0,
 'HalfBath': 0,
 'BedroomAbvGr': 24,
 'KitchenAbvGr': 41,
 'TotRmsAbvGrd': 16,
 'Fireplaces': 1,
 'GarageCars': 3,
 'GarageArea': 14,
 'WoodDeckSF': 16,
 'OpenPorchSF': 53,
 'EnclosedPorch': 128,
 '3SsnPorch': 15,
 'ScreenPorch': 67,
 'PoolArea': 3,
 'MiscVal': 35,
 'MoSold': 0,
 'YrSold': 0}

In [113]:
def log_transform(df, columns):
    for column in columns:
        df[column] = np.log1p(df[column])
    return df

# Function to replace outliers with mean
def replace_outliers_with_mean(df, col):
    q1 = np.percentile(df[col], 25)
    q3 = np.percentile(df[col], 75)
    iqr = q3 - q1
    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr
    mean_value = df[col].mean()
    
    df[col] = np.where((df[col] < lower_bound) | (df[col] > upper_bound), mean_value, df[col])

# Apply appropriate outlier handling for each column
for col, num_outliers in outlier_count.items():
    if num_outliers < 30:
        replace_outliers_with_mean(df1, col)
    else:
        log_transform(df1, [col])

# Verify changes
print("Outliers handled successfully.")

Outliers handled successfully.


In [114]:
df1


Id  MSSubClass MSZoning  LotFrontage   LotArea Street Alley LotShape  \
0     142.0    3.044522       RL         78.0  9.362718   Pave    no      Reg   
1     365.0    4.110874       RL         70.0  9.841665   Pave    no      IR1   
2     799.0    4.110874       RL        104.0  9.511851   Pave    no      Reg   
3     976.0    5.081404       FV         70.0  7.883069   Pave    no      Reg   
4    1158.0    4.795791       RL         34.0  8.517593   Pave    no      IR1   
..      ...         ...      ...          ...       ...    ...   ...      ...   
929   782.0    4.110874       RL         65.0  8.875427   Pave    no      Reg   
930   614.0    3.044522       RL         70.0  9.036344   Pave    no      Reg   
931   878.0    4.110874       RL         74.0  9.086476   Pave    no      Reg   
932  1037.0    3.044522       RL         89.0  9.464905   Pave    no      IR1   
933   873.0    3.044522       RL         74.0  9.093020   Pave    no      Reg   

    LandContour Utilities  ... ScreenPorch PoolArea PoolQC Fence MiscFeature  \
0           Lvl    AllPub  ...         0.0      0.0     no    no          no   
1           Lvl    AllPub  ...         0.0      0.0     no    no          no   
2           Lvl    AllPub  ...         0.0      0.0     no    no          no   
3           Lvl    AllPub  ...         0.0      0.0     no    no          no   
4           Lvl    AllPub  ...         0.0      0.0     no    no          no   
..          ...       ...  ...         ...      ...    ...   ...         ...   
929         Lvl    AllPub  ...         0.0      0.0     no    no          no   
930         Lvl    AllPub  ...         0.0      0.0     no    no          no   
931         Lvl    AllPub  ...         0.0      0.0     no    no          no   
932         HLS    AllPub  ...         0.0      0.0     no    no          no   
933         Lvl    AllPub  ...         0.0      0.0     no  GdWo          no   

    MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0.0    1.0  2006.0        WD         Normal  
1       0.0    7.0  2006.0        WD         Normal  
2       0.0    7.0  2009.0       New        Partial  
3       0.0    4.0  2006.0        WD         Normal  
4       0.0    7.0  2009.0        WD         Normal  
..      ...    ...     ...       ...            ...  
929     0.0    7.0  2006.0        WD         Normal  
930     0.0   12.0  2007.0       New        Partial  
931     0.0    6.0  2009.0        WD         Normal  
932     0.0    9.0  2009.0        WD         Normal  
933     0.0   10.0  2008.0        WD         Normal  

[934 rows x 80 columns]

# outlier for test data

In [115]:
def count_outliers(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = df[(df[column] < lower_bound) | (df[column] > upper_bound)]
    return len(outliers)
outlier_count = {}
for column in df2.select_dtypes(include='int'):
    outlier_count[column] = count_outliers(df2, column)
for column, count in outlier_count.items():
    print(f'{column}: {count} outliers')

Id: 0 outliers
MSSubClass: 19 outliers
LotArea: 19 outliers
OverallQual: 0 outliers
OverallCond: 20 outliers
YearBuilt: 0 outliers
YearRemodAdd: 0 outliers
BsmtFinSF1: 1 outliers
BsmtFinSF2: 28 outliers
BsmtUnfSF: 2 outliers
TotalBsmtSF: 12 outliers
1stFlrSF: 4 outliers
2ndFlrSF: 0 outliers
LowQualFinSF: 6 outliers
GrLivArea: 5 outliers
BsmtFullBath: 0 outliers
BsmtHalfBath: 14 outliers
FullBath: 0 outliers
HalfBath: 0 outliers
BedroomAbvGr: 4 outliers
KitchenAbvGr: 13 outliers
TotRmsAbvGrd: 8 outliers
Fireplaces: 3 outliers
GarageCars: 0 outliers
GarageArea: 3 outliers
WoodDeckSF: 8 outliers
OpenPorchSF: 11 outliers
EnclosedPorch: 30 outliers
3SsnPorch: 8 outliers
ScreenPorch: 21 outliers
PoolArea: 2 outliers
MiscVal: 8 outliers
MoSold: 0 outliers
YrSold: 0 outliers


In [116]:
def log_transform(df, columns):
    for column in columns:
        df[column] = np.log1p(df[column])
    return df

# Function to replace outliers with mean
def replace_outliers_with_mean(df, col):
    q1 = np.percentile(df[col], 25)
    q3 = np.percentile(df[col], 75)
    iqr = q3 - q1
    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr
    mean_value = df[col].mean()
    
    df[col] = np.where((df[col] < lower_bound) | (df[col] > upper_bound), mean_value, df[col])

# Apply appropriate outlier handling for each column
for col, num_outliers in outlier_count.items():
    if num_outliers < 30:
        replace_outliers_with_mean(df2, col)
    else:
        log_transform(df2, [col])

# Verify changes
print("Outliers handled successfully.")

Outliers handled successfully.


In [117]:
df1.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 934 entries, 0 to 933
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             934 non-null    float64
 1   MSSubClass     934 non-null    float64
 2   MSZoning       934 non-null    object 
 3   LotFrontage    934 non-null    float64
 4   LotArea        934 non-null    float64
 5   Street         934 non-null    object 
 6   Alley          934 non-null    object 
 7   LotShape       934 non-null    object 
 8   LandContour    934 non-null    object 
 9   Utilities      934 non-null    object 
 10  LotConfig      934 non-null    object 
 11  LandSlope      934 non-null    object 
 12  Neighborhood   934 non-null    object 
 13  Condition1     934 non-null    object 
 14  Condition2     934 non-null    object 
 15  BldgType       934 non-null    object 
 16  HouseStyle     934 non-null    object 
 17  OverallQual    934 non-null    float64
 18  OverallCon

In [118]:
df1['Alley'].value_counts()

Alley
no      878
Grvl     29
Pave     27
Name: count, dtype: int64

# tha corr after solve the null and outlier

In [119]:
df1=pd.concat([df1, z], axis=1)

In [120]:
df1

Id  MSSubClass MSZoning  LotFrontage   LotArea Street Alley LotShape  \
0     142.0    3.044522       RL         78.0  9.362718   Pave    no      Reg   
1     365.0    4.110874       RL         70.0  9.841665   Pave    no      IR1   
2     799.0    4.110874       RL        104.0  9.511851   Pave    no      Reg   
3     976.0    5.081404       FV         70.0  7.883069   Pave    no      Reg   
4    1158.0    4.795791       RL         34.0  8.517593   Pave    no      IR1   
..      ...         ...      ...          ...       ...    ...   ...      ...   
929   782.0    4.110874       RL         65.0  8.875427   Pave    no      Reg   
930   614.0    3.044522       RL         70.0  9.036344   Pave    no      Reg   
931   878.0    4.110874       RL         74.0  9.086476   Pave    no      Reg   
932  1037.0    3.044522       RL         89.0  9.464905   Pave    no      IR1   
933   873.0    3.044522       RL         74.0  9.093020   Pave    no      Reg   

    LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal  \
0           Lvl    AllPub  ...      0.0     no    no          no     0.0   
1           Lvl    AllPub  ...      0.0     no    no          no     0.0   
2           Lvl    AllPub  ...      0.0     no    no          no     0.0   
3           Lvl    AllPub  ...      0.0     no    no          no     0.0   
4           Lvl    AllPub  ...      0.0     no    no          no     0.0   
..          ...       ...  ...      ...    ...   ...         ...     ...   
929         Lvl    AllPub  ...      0.0     no    no          no     0.0   
930         Lvl    AllPub  ...      0.0     no    no          no     0.0   
931         Lvl    AllPub  ...      0.0     no    no          no     0.0   
932         HLS    AllPub  ...      0.0     no    no          no     0.0   
933         Lvl    AllPub  ...      0.0     no  GdWo          no     0.0   

    MoSold  YrSold  SaleType  SaleCondition  SalePrice  
0      1.0  2006.0        WD         Normal     260000  
1      7.0  2006.0        WD         Normal     190000  
2      7.0  2009.0       New        Partial     485000  
3      4.0  2006.0        WD         Normal     165000  
4      7.0  2009.0        WD         Normal     230000  
..     ...     ...       ...            ...        ...  
929    7.0  2006.0        WD         Normal     175900  
930   12.0  2007.0       New        Partial     147000  
931    6.0  2009.0        WD         Normal     350000  
932    9.0  2009.0        WD         Normal     315500  
933   10.0  2008.0        WD         Normal     116000  

[934 rows x 81 columns]

In [121]:

# Correlation matrix
correlation_matrix = numeric_data.corr()

# Extract the correlations with SalePrice and sort them
correlation_with_saleprice = correlation_matrix['SalePrice'].sort_values(ascending=False)

print(correlation_with_saleprice)

SalePrice        1.000000
OverallQual      0.777172
GrLivArea        0.720202
TotalBsmtSF      0.658005
1stFlrSF         0.643107
GarageCars       0.633413
GarageArea       0.617788
FullBath         0.547057
TotRmsAbvGrd     0.522824
YearBuilt        0.506217
YearRemodAdd     0.486246
MasVnrArea       0.473106
Fireplaces       0.472682
GarageYrBlt      0.445172
BsmtFinSF1       0.414819
LotFrontage      0.389367
OpenPorchSF      0.303031
2ndFlrSF         0.297547
WoodDeckSF       0.289732
LotArea          0.263302
HalfBath         0.246890
BsmtUnfSF        0.225948
BsmtFullBath     0.221758
BedroomAbvGr     0.157828
PoolArea         0.156577
ScreenPorch      0.102507
3SsnPorch        0.054778
MoSold           0.034923
Id               0.007412
BsmtHalfBath    -0.004336
BsmtFinSF2      -0.015181
LowQualFinSF    -0.015383
YrSold          -0.019660
MiscVal         -0.023140
OverallCond     -0.075046
MSSubClass      -0.091493
EnclosedPorch   -0.118967
KitchenAbvGr    -0.134543
Name: SalePr

In [122]:
target_feature = 'SalePrice' 
correlations = correlation_ratios(df1,cat_feat,target_feature)
sorted_correlations = sorted(correlations.items(), key=lambda item: item[1], reverse=True)

for cat_feature, ratio in sorted_correlations:
    print(f"{cat_feature}: {ratio}")


Neighborhood: 0.7241911234293448
BsmtQual: 0.684105342658258
KitchenQual: 0.6783575663504264
ExterQual: 0.6745851410603996
FireplaceQu: 0.5404908503427579
GarageFinish: 0.5279356661831272
Foundation: 0.5095458313146748
BsmtFinType1: 0.47777567059937304
GarageType: 0.46613798102358345
MasVnrType: 0.44345322810552545
HeatingQC: 0.4384719014992851
Exterior2nd: 0.40788260221638784
BsmtExposure: 0.3942617808531408
Exterior1st: 0.374763698717642
SaleType: 0.3631197306525389
SaleCondition: 0.34906217315143157
MSZoning: 0.30892450606188226
GarageCond: 0.27799219312641793
GarageQual: 0.2744025570918595
RoofStyle: 0.26435679820973385
HouseStyle: 0.2633784396100664
LotShape: 0.263170892723789
CentralAir: 0.2584665999076131
RoofMatl: 0.23915735336343386
Electrical: 0.2382116297729705
BsmtCond: 0.2371410258237414
PavedDrive: 0.22467358813568108
BsmtFinType2: 0.1888919125870447
Fence: 0.18160868644077655
PoolQC: 0.18049019186823387
Condition1: 0.18025665347156986
BldgType: 0.17420960653160872
ExterC

note
1. try the model before we clean the outlier
2.reduce to 0.15

***just for test***

In [123]:
df1 = df1.drop(columns=['BldgType'])
df1 = df1.drop(columns=['RoofMatl'])
df1 = df1.drop(columns=['Alley'])
df1 = df1.drop(columns=['LandContour'])
df1 = df1.drop(columns=['LotConfig'])
df1 = df1.drop(columns=['Functional'])
df1 = df1.drop(columns=['PoolQC'])
df1 = df1.drop(columns=['Condition2'])
df1 = df1.drop(columns=['MiscFeature'])
df1 = df1.drop(columns=['Street'])
df1 = df1.drop(columns=['LandSlope'])
df1 = df1.drop(columns=['Utilities'])
df1 = df1.drop(columns=['BsmtUnfSF'])
df1 = df1.drop(columns=['PoolArea'])
df1 = df1.drop(columns=['ScreenPorch'])
df1 = df1.drop(columns=['3SsnPorch'])
df1 = df1.drop(columns=['MoSold'])
df1 = df1.drop(columns=['OverallCond'])
df1 = df1.drop(columns=['BsmtHalfBath'])
df1 = df1.drop(columns=['BsmtFinSF2'])
df1 = df1.drop(columns=['MSSubClass'])
df1 = df1.drop(columns=['YrSold'])
df1 = df1.drop(columns=['LowQualFinSF'])
df1 = df1.drop(columns=['MiscVal'])
df1 = df1.drop(columns=['KitchenAbvGr'])
df1 = df1.drop(columns=['Id'])

In [124]:
df2 = df2.drop(columns=['BldgType'])
df2 = df2.drop(columns=['RoofMatl'])
df2 = df2.drop(columns=['Alley'])
df2 = df2.drop(columns=['LandContour'])
df2 = df2.drop(columns=['LotConfig'])
df2 = df2.drop(columns=['Functional'])
df2 = df2.drop(columns=['PoolQC'])
df2 = df2.drop(columns=['Condition2'])
df2 = df2.drop(columns=['MiscFeature'])
df2 = df2.drop(columns=['Street'])
df2 = df2.drop(columns=['LandSlope'])
df2 = df2.drop(columns=['Utilities'])
df2 = df2.drop(columns=['BsmtUnfSF'])
df2 = df2.drop(columns=['PoolArea'])
df2 = df2.drop(columns=['ScreenPorch'])
df2 = df2.drop(columns=['3SsnPorch'])
df2 = df2.drop(columns=['MoSold'])
df2 = df2.drop(columns=['OverallCond'])
df2 = df2.drop(columns=['BsmtHalfBath'])
df2 = df2.drop(columns=['BsmtFinSF2'])
df2 = df2.drop(columns=['MSSubClass'])
df2 = df2.drop(columns=['YrSold'])
df2 = df2.drop(columns=['LowQualFinSF'])
df2 = df2.drop(columns=['MiscVal'])
df2 = df2.drop(columns=['KitchenAbvGr'])


In [125]:
df1.shape

(934, 55)

# encoding

In [126]:
cat_feat = df1.select_dtypes(include=['object']).columns
nue_feat = df1.select_dtypes(exclude=['object']).columns
# Label encoding for binary categorical columns
label_encoder = LabelEncoder()
binary_cols = [col for col in cat_feat if df1[col].nunique() == 2]

for col in binary_cols:
    df1[col] = label_encoder.fit_transform(df1[col])

# One-hot encoding for remaining categorical columns
one_hot_cols = [col for col in cat_feat if col not in binary_cols]

df1 = pd.get_dummies(df1, columns=one_hot_cols, drop_first=True)


In [127]:
df1

LotFrontage   LotArea  OverallQual  YearBuilt  YearRemodAdd  MasVnrArea  \
0           78.0  9.362718          7.0     2005.0        2005.0         0.0   
1           70.0  9.841665          6.0     1976.0        1976.0       120.0   
2          104.0  9.511851          9.0     2008.0        2009.0       860.0   
3           70.0  7.883069          7.0     2000.0        2000.0         0.0   
4           34.0  8.517593          7.0     2007.0        2008.0       166.0   
..           ...       ...          ...        ...           ...         ...   
929         65.0  8.875427          6.0     1992.0        1992.0         0.0   
930         70.0  9.036344          5.0     2007.0        2007.0         0.0   
931         74.0  9.086476          9.0     2004.0        2005.0       216.0   
932         89.0  9.464905          9.0     2007.0        2008.0        70.0   
933         74.0  9.093020          5.0     1953.0        1996.0         0.0   

     BsmtFinSF1  TotalBsmtSF  CentralAir  1stFlrSF  ...  SaleType_ConLI  \
0        1300.0     7.458763           1    1734.0  ...           False   
1         712.0     6.680855           1     790.0  ...           False   
2           0.0     7.563720           1    1966.0  ...           False   
3         641.0     6.513230           1     673.0  ...           False   
4         904.0     7.181592           1    1314.0  ...           False   
..          ...          ...         ...       ...  ...             ...   
929       387.0     6.635947           1     810.0  ...           False   
930       206.0     7.021976           1    1120.0  ...           False   
931      1170.0     7.288244           1    1462.0  ...           False   
932      1022.0     7.390799           1    1620.0  ...           False   
933         0.0     4.663439           1     910.0  ...           False   

     SaleType_ConLw  SaleType_New  SaleType_Oth  SaleType_WD  \
0             False         False         False         True   
1             False         False         False         True   
2             False          True         False        False   
3             False         False         False         True   
4             False         False         False         True   
..              ...           ...           ...          ...   
929           False         False         False         True   
930           False          True         False        False   
931           False         False         False         True   
932           False         False         False         True   
933           False         False         False         True   

     SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
0                    False                 False                 False   
1                    False                 False                 False   
2                    False                 False                 False   
3                    False                 False                 False   
4                    False                 False                 False   
..                     ...                   ...                   ...   
929                  False                 False                 False   
930                  False                 False                 False   
931                  False                 False                 False   
932                  False                 False                 False   
933                  False                 False                 False   

     SaleCondition_Normal  SaleCondition_Partial  
0                    True                  False  
1                    True                  False  
2                   False                   True  
3                    True                  False  
4                    True                  False  
..                    ...                    ...  
929                  True                  False  
930                 False                   True  
931                  True 

In [128]:
cat_feat = df2.select_dtypes(include=['object']).columns
nue_feat = df2.select_dtypes(exclude=['object']).columns
# Label encoding for binary categorical columns
label_encoder = LabelEncoder()
binary_cols = [col for col in cat_feat if df2[col].nunique() == 2]

for col in binary_cols:
    df2[col] = label_encoder.fit_transform(df2[col])

# One-hot encoding for remaining categorical columns
one_hot_cols = [col for col in cat_feat if col not in binary_cols]

df2 = pd.get_dummies(df2, columns=one_hot_cols, drop_first=True)


In [129]:
df2

Id  LotFrontage  LotArea  OverallQual  YearBuilt  YearRemodAdd  \
0      42.0        115.0  16905.0          5.0     1959.0        1959.0   
1    1156.0         90.0  10768.0          5.0     1976.0        2004.0   
2     239.0         93.0  12030.0          8.0     2007.0        2007.0   
3     810.0         90.0   8100.0          5.0     1898.0        1965.0   
4     471.0         65.0   6820.0          8.0     1985.0        1985.0   
..      ...          ...      ...          ...        ...           ...   
229   389.0         93.0   9382.0          7.0     1999.0        2000.0   
230  1193.0         60.0   9600.0          5.0     1925.0        1994.0   
231    95.0         69.0   9337.0          6.0     1997.0        1997.0   
232  1171.0         76.0   9880.0          6.0     1977.0        1977.0   
233   190.0         41.0   4923.0          8.0     2001.0        2002.0   

     MasVnrArea  BsmtFinSF1  TotalBsmtSF  CentralAir  ...  SaleType_ConLD  \
0           0.0       967.0       1350.0           1  ...           False   
1           0.0      1157.0       1437.0           1  ...           False   
2         254.0         0.0       1694.0           1  ...           False   
3           0.0         0.0        849.0           0  ...           False   
4           0.0       368.0       1488.0           1  ...           False   
..          ...         ...          ...         ...  ...             ...   
229       125.0         0.0       1468.0           1  ...           False   
230         0.0         0.0        702.0           1  ...           False   
231         0.0       648.0        824.0           1  ...           False   
232         0.0       522.0       1096.0           1  ...           False   
233         0.0      1153.0       1593.0           1  ...           False   

     SaleType_ConLI  SaleType_New  SaleType_Oth  SaleType_WD  \
0             False         False         False         True   
1             False         False         False         True   
2             False          True         False        False   
3             False         False         False         True   
4             False         False         False         True   
..              ...           ...           ...          ...   
229           False         False         False         True   
230           False         False         False         True   
231           False         False         False         True   
232           False         False         False         True   
233           False         False         False         True   

     SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
0                    False                 False                 False   
1                    False                 False                 False   
2                    False                 False                 False   
3                    False                 False                 False   
4                    False                 False                 False   
..                     ...                   ...                   ...   
229                  False                 False                 False   
230                  False                 False                 False   
231                  False                 False                 False   
232                  False                 False                 False   
233                  False                 False                 False   

     SaleCondition_Normal  SaleCondition_Partial  
0                    True                  False  
1                    True                  False  
2                   False                   True  
3                    True                  False  
4                    True                  False  
..                    ...                    ...  
229                  True                  False  
230                  True                  False  
231                  True                  False  
232            

In [130]:
df1.columns

Index(['LotFrontage', 'LotArea', 'OverallQual', 'YearBuilt', 'YearRemodAdd',
       'MasVnrArea', 'BsmtFinSF1', 'TotalBsmtSF', 'CentralAir', '1stFlrSF',
       ...
       'SaleType_ConLI', 'SaleType_ConLw', 'SaleType_New', 'SaleType_Oth',
       'SaleType_WD', 'SaleCondition_AdjLand', 'SaleCondition_Alloca',
       'SaleCondition_Family', 'SaleCondition_Normal',
       'SaleCondition_Partial'],
      dtype='object', length=197)

In [131]:
df1.shape

(934, 197)

In [132]:
y = df1['SalePrice']

In [133]:
for col in df1.columns:
    if col not in df2.columns:
        df1.drop(col,axis=1,inplace=True)
        

In [134]:
ids = df2['Id']
for col in df2.columns:
    if col not in df1.columns:
        df2.drop(col,axis=1,inplace=True)
       

In [135]:
ids.dtype

dtype('float64')

In [136]:
df1.shape

(934, 175)

In [137]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Separate the target variable and features
X = df1


# Initialize the scaler
scaler = StandardScaler()

# Fit and transform the training data
X = scaler.fit_transform(X)
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)



# Display the shape of the datasets
X_train.shape, X_test.shape, y_train.shape, y_test.shape


((747, 175), (187, 175), (747,), (187,))

In [138]:
df1.shape

(934, 175)

In [139]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Initialize the Linear Regression model
model = LinearRegression()

# Fit the model to the training data
model.fit(X_train, y_train)

# Predict on the test data
y_pred = model.predict(X_test)

# Evaluate the model
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = mse ** 0.5
r2 = r2_score(y_test, y_pred)

print(f'Mean Absolute Error (MAE): {mae}')
print(f'Mean Squared Error (MSE): {mse}')
print(f'Root Mean Squared Error (RMSE): {rmse}')
print(f'R-squared (R2): {r2}')


Mean Absolute Error (MAE): 1231060216853081.8
Mean Squared Error (MSE): 2.834002311443106e+32
Root Mean Squared Error (RMSE): 1.6834495274415286e+16
R-squared (R2): -3.842015784806086e+22


In [140]:
# Importing necessary libraries
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.neural_network import MLPRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score


# Random Forest Regressor
rf_regressor = RandomForestRegressor()
rf_regressor.fit(X_train, y_train)
rf_preds = rf_regressor.predict(X_test)
rf_mae = mean_absolute_error(y_test, rf_preds)
rf_mse = mean_squared_error(y_test, rf_preds)
rf_r2 = r2_score(y_test, rf_preds)
rsq=np.sqrt(rf_mse)

# Gradient Boosting Regressor
gb_regressor = GradientBoostingRegressor()
gb_regressor.fit(X_train, y_train)
gb_preds = gb_regressor.predict(X_test)
gb_mae = mean_absolute_error(y_test, gb_preds)
gb_mse = mean_squared_error(y_test, gb_preds)
gb_r2 = r2_score(y_test, gb_preds)

# Support Vector Regressor
svr_regressor = SVR()
svr_regressor.fit(X_train, y_train)
svr_preds = svr_regressor.predict(X_test)
svr_mae = mean_absolute_error(y_test, svr_preds)
svr_mse = mean_squared_error(y_test, svr_preds)
svr_r2 = r2_score(y_test, svr_preds)

# Neural Network Regressor
nn_regressor = MLPRegressor()
nn_regressor.fit(X_train, y_train)
nn_preds = nn_regressor.predict(X_test)
nn_mae = mean_absolute_error(y_test, nn_preds)
nn_mse = mean_squared_error(y_test, nn_preds)
nn_r2 = r2_score(y_test, nn_preds)

# K-Nearest Neighbors Regressor
knn_regressor = KNeighborsRegressor()
knn_regressor.fit(X_train, y_train)
knn_preds = knn_regressor.predict(X_test)
knn_mae = mean_absolute_error(y_test, knn_preds)
knn_mse = mean_squared_error(y_test, knn_preds)
knn_r2 = r2_score(y_test, knn_preds)

# Print the results
print("Random Forest Regressor:")
print("MAE:", rf_mae)
print("MSE:", rf_mse)
print("R-squared:", rf_r2)
print("\nGradient Boosting Regressor:")
print("MAE:", gb_mae)
print("MSE:", gb_mse)
print("R-squared:", gb_r2)
print("\nSupport Vector Regressor:")
print("MAE:", svr_mae)
print("MSE:", svr_mse)
print("R-squared:", svr_r2)
print("\nNeural Network Regressor:")
print("MAE:", nn_mae)
print("MSE:", nn_mse)
print("R-squared:", nn_r2)
print("\nK-Nearest Neighbors Regressor:")
print("MAE:", knn_mae)
print("MSE:", knn_mse)
print("R-squared:", knn_r2)
print(rsq)
'''
MAE: 20212.62
MSE: 1058105613.97
R-squared: 0.8566
Gradient Boosting Regressor:

MAE: 19774.58
MSE: 1039868474.63
R-squared: 0.8590

'''

Random Forest Regressor:
MAE: 19743.30962566845
MSE: 965842115.2543668
R-squared: 0.8690621868071728

Gradient Boosting Regressor:
MAE: 20169.07631918002
MSE: 1082008904.5555832
R-squared: 0.8533136238521114

Support Vector Regressor:
MAE: 61259.18498727267
MSE: 7953425793.248914
R-squared: -0.07823438666802374

Neural Network Regressor:
MAE: 185975.8427514455
MSE: 41698863150.243744
R-squared: -4.653054331847942

K-Nearest Neighbors Regressor:
MAE: 30364.19465240642
MSE: 1927854636.673797
R-squared: 0.7386435460900991
31078.00050283748


/opt/conda/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


'\nMAE: 20212.62\nMSE: 1058105613.97\nR-squared: 0.8566\nGradient Boosting Regressor:\n\nMAE: 19774.58\nMSE: 1039868474.63\nR-squared:\xa00.8590\n\n'

In [141]:
df1.describe()

LotFrontage     LotArea  OverallQual    YearBuilt  YearRemodAdd  \
count   934.000000  934.000000   934.000000   934.000000    934.000000   
mean     69.657388    9.101285     6.150191  1973.265503   1985.350107   
std      20.639725    0.522143     1.365228    28.949793     20.480156   
min      21.000000    7.170888     2.000000  1890.000000   1950.000000   
25%      60.000000    8.922792     5.000000  1955.250000   1968.000000   
50%      70.000000    9.164610     6.000000  1976.000000   1994.500000   
75%      79.000000    9.362675     7.000000  2001.000000   2004.000000   
max     182.000000   12.279537    10.000000  2010.000000   2010.000000   

        MasVnrArea   BsmtFinSF1  TotalBsmtSF  CentralAir     1stFlrSF  ...  \
count   934.000000   934.000000   934.000000  934.000000   934.000000  ...   
mean    106.965739   450.462623     6.741810    0.935760  1147.387890  ...   
std     184.149861   430.758583     1.195978    0.245311   342.811063  ...   
min       0.000000     0.000000     0.000000    0.000000   334.000000  ...   
25%       0.000000     0.000000     6.680855    1.000000   882.500000  ...   
50%       0.000000   403.000000     6.911249    1.000000  1085.500000  ...   
75%     169.750000   732.000000     7.172425    1.000000  1370.000000  ...   
max    1600.000000  1810.000000     8.073091    1.000000  2129.000000  ...   

         HalfBath  BedroomAbvGr  TotRmsAbvGrd  Fireplaces  GarageYrBlt  \
count  934.000000    934.000000    934.000000  934.000000   934.000000   
mean     0.384368      2.838543      6.432805    0.608792  1979.604925   
std      0.506138      0.700619      1.443204    0.624790    23.503591   
min      0.000000      1.000000      2.000000    0.000000  1900.000000   
25%      0.000000      2.000000      5.000000    0.000000  1964.000000   
50%      0.000000      3.000000      6.000000    1.000000  1981.500000   
75%      1.000000      3.000000      7.000000    1.000000  2001.000000   
max      2.000000      4.000000     10.000000    2.000000  2010.000000   

       GarageCars  GarageArea  WoodDeckSF  OpenPorchSF  EnclosedPorch  
count  934.000000  934.000000  934.000000   934.000000     934.000000  
mean     1.779830  467.386993   91.530882     2.369888       0.671016  
std      0.733417  197.947105  108.991665     2.150614       1.699135  
min      0.000000    0.000000    0.000000     0.000000       0.000000  
25%      1.000000  336.000000    0.000000     0.000000       0.000000  
50%      2.000000  476.817987   29.000000     3.332205       0.000000  
75%      2.000000  576.000000  168.000000     4.262680       0.000000  
max      3.000000  928.000000  431.000000     6.306275       6.315358  

[8 rows x 24 columns]

In [142]:

X = df2
scaler = StandardScaler()
X =pd.DataFrame(scaler.fit_transform(X))


In [143]:
df1.head()

LotFrontage   LotArea  OverallQual  YearBuilt  YearRemodAdd  MasVnrArea  \
0         78.0  9.362718          7.0     2005.0        2005.0         0.0   
1         70.0  9.841665          6.0     1976.0        1976.0       120.0   
2        104.0  9.511851          9.0     2008.0        2009.0       860.0   
3         70.0  7.883069          7.0     2000.0        2000.0         0.0   
4         34.0  8.517593          7.0     2007.0        2008.0       166.0   

   BsmtFinSF1  TotalBsmtSF  CentralAir  1stFlrSF  ...  SaleType_ConLD  \
0      1300.0     7.458763           1    1734.0  ...           False   
1       712.0     6.680855           1     790.0  ...           False   
2         0.0     7.563720           1    1966.0  ...           False   
3       641.0     6.513230           1     673.0  ...           False   
4       904.0     7.181592           1    1314.0  ...           False   

   SaleType_ConLI  SaleType_New  SaleType_Oth  SaleType_WD  \
0           False         False         False         True   
1           False         False         False         True   
2           False          True         False        False   
3           False         False         False         True   
4           False         False         False         True   

   SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
0                  False                 False                 False   
1                  False                 False                 False   
2                  False                 False                 False   
3                  False                 False                 False   
4                  False                 False                 False   

   SaleCondition_Normal  SaleCondition_Partial  
0                  True                  False  
1                  True                  False  
2                 False                   True  
3                  True                  False  
4                  True                  False  

[5 rows x 175 columns]

In [144]:
df2.head()

LotFrontage  LotArea  OverallQual  YearBuilt  YearRemodAdd  MasVnrArea  \
0        115.0  16905.0          5.0     1959.0        1959.0         0.0   
1         90.0  10768.0          5.0     1976.0        2004.0         0.0   
2         93.0  12030.0          8.0     2007.0        2007.0       254.0   
3         90.0   8100.0          5.0     1898.0        1965.0         0.0   
4         65.0   6820.0          8.0     1985.0        1985.0         0.0   

   BsmtFinSF1  TotalBsmtSF  CentralAir  1stFlrSF  ...  SaleType_ConLD  \
0       967.0       1350.0           1    1328.0  ...           False   
1      1157.0       1437.0           1    1437.0  ...           False   
2         0.0       1694.0           1    1694.0  ...           False   
3         0.0        849.0           0    1075.0  ...           False   
4       368.0       1488.0           1    1502.0  ...           False   

   SaleType_ConLI  SaleType_New  SaleType_Oth  SaleType_WD  \
0           False         False         False         True   
1           False         False         False         True   
2           False          True         False        False   
3           False         False         False         True   
4           False         False         False         True   

   SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
0                  False                 False                 False   
1                  False                 False                 False   
2                  False                 False                 False   
3                  False                 False                 False   
4                  False                 False                 False   

   SaleCondition_Normal  SaleCondition_Partial  
0                  True                  False  
1                  True                  False  
2                 False                   True  
3                  True                  False  
4                  True                  False  

[5 rows x 175 columns]

In [145]:
type(df2)

pandas.core.frame.DataFrame

In [146]:
df2.columns


Index(['LotFrontage', 'LotArea', 'OverallQual', 'YearBuilt', 'YearRemodAdd',
       'MasVnrArea', 'BsmtFinSF1', 'TotalBsmtSF', 'CentralAir', '1stFlrSF',
       ...
       'SaleType_ConLD', 'SaleType_ConLI', 'SaleType_New', 'SaleType_Oth',
       'SaleType_WD', 'SaleCondition_AdjLand', 'SaleCondition_Alloca',
       'SaleCondition_Family', 'SaleCondition_Normal',
       'SaleCondition_Partial'],
      dtype='object', length=175)

In [147]:
df1.shape

(934, 175)

In [148]:
df2.head()


LotFrontage  LotArea  OverallQual  YearBuilt  YearRemodAdd  MasVnrArea  \
0        115.0  16905.0          5.0     1959.0        1959.0         0.0   
1         90.0  10768.0          5.0     1976.0        2004.0         0.0   
2         93.0  12030.0          8.0     2007.0        2007.0       254.0   
3         90.0   8100.0          5.0     1898.0        1965.0         0.0   
4         65.0   6820.0          8.0     1985.0        1985.0         0.0   

   BsmtFinSF1  TotalBsmtSF  CentralAir  1stFlrSF  ...  SaleType_ConLD  \
0       967.0       1350.0           1    1328.0  ...           False   
1      1157.0       1437.0           1    1437.0  ...           False   
2         0.0       1694.0           1    1694.0  ...           False   
3         0.0        849.0           0    1075.0  ...           False   
4       368.0       1488.0           1    1502.0  ...           False   

   SaleType_ConLI  SaleType_New  SaleType_Oth  SaleType_WD  \
0           False         False         False         True   
1           False         False         False         True   
2           False          True         False        False   
3           False         False         False         True   
4           False         False         False         True   

   SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
0                  False                 False                 False   
1                  False                 False                 False   
2                  False                 False                 False   
3                  False                 False                 False   
4                  False                 False                 False   

   SaleCondition_Normal  SaleCondition_Partial  
0                  True                  False  
1                  True                  False  
2                 False                   True  
3                  True                  False  
4                  True                  False  

[5 rows x 175 columns]

In [154]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor



X_train = df1
y_train = y

# Initialize and train the RandomForestRegressor model
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

X_test = X

# Make predictions on the test data
predictions = model.predict(X_test)

# Create a dataframe with IDs and predictions
results = pd.DataFrame({'Id': ids, 'SalePrice': predictions})
results['Id'] = pd.to_numeric(results['Id'], errors='coerce')

# Step 2: Convert the column to Int32
results['Id'] = results['Id'].astype('Int32')
# Save the results to a CSV file
results.to_csv('predictions.csv', index=False)

print("Predictions saved to predictions.csv")

Predictions saved to predictions.csv


/opt/conda/lib/python3.10/site-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(


In [155]:
z=pd.read_csv('/kaggle/working/predictions.csv')
z

Id  SalePrice
0      42    61591.0
1    1156    63164.5
2     239    70752.5
3     810    62127.0
4     471    62490.5
..    ...        ...
229   389    59289.0
230  1193    62801.5
231    95    58072.0
232  1171    62828.0
233   190    61251.5

[234 rows x 2 columns]

In [156]:
z['Id']

0        42
1      1156
2       239
3       810
4       471
       ... 
229     389
230    1193
231      95
232    1171
233     190
Name: Id, Length: 234, dtype: int64

In [157]:
'''import pandas as pd
from sklearn.ensemble import RandomForestRegressor

# Assume df1 and df2 are already defined

X_train = df1.drop(columns=['SalePrice', 'Id'])
y_train = df1['SalePrice']

# Initialize and train the RandomForestRegressor model
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# Align columns of df2 with X_train
X_test = df2[X_train.columns]

# Make predictions on the test data
predictions = model.predict(X_test)

# Extract 'Id' column from df2
ids = df2['Id']

# Create a dataframe with IDs and predictions
results = pd.DataFrame({'Id': ids, 'predictions': predictions})

# Save the results to a CSV file
results.to_csv('predictions.csv', index=False)

print("Predictions saved to predictions.csv")
'''

'import pandas as pd\nfrom sklearn.ensemble import RandomForestRegressor\n\n# Assume df1 and df2 are already defined\n\nX_train = df1.drop(columns=[\'SalePrice\', \'Id\'])\ny_train = df1[\'SalePrice\']\n\n# Initialize and train the RandomForestRegressor model\nmodel = RandomForestRegressor(n_estimators=100, random_state=42)\nmodel.fit(X_train, y_train)\n\n# Align columns of df2 with X_train\nX_test = df2[X_train.columns]\n\n# Make predictions on the test data\npredictions = model.predict(X_test)\n\n# Extract \'Id\' column from df2\nids = df2[\'Id\']\n\n# Create a dataframe with IDs and predictions\nresults = pd.DataFrame({\'Id\': ids, \'predictions\': predictions})\n\n# Save the results to a CSV file\nresults.to_csv(\'predictions.csv\', index=False)\n\nprint("Predictions saved to predictions.csv")\n'